src/middleware/call.rs

```diff
@@ -1,39 +1,32 @@
-use std::sync::Arc;
-
-use async_trait::async_trait;
-use jsonrpsee::{
-    core::{Error, JsonValue},
-    types::Params,
-};
-
 use crate::client::Client;
+use async_trait::async_trait;
+use jsonrpsee::core::{Error, JsonValue};
+use std::sync::Arc;
 
-use super::{Middleware, NextFn};
-
-pub struct CallRequest {
-    pub method: String,
-    pub params: Params<'static>,
-}
+use super::{Middleware, NextFn, Request};
 
-pub struct UpstreamMiddleware {
+pub struct CallMiddleware {
     client: Arc<Client>,
 }
 
-impl UpstreamMiddleware {
-    pub fn new(client: Arc<Client>) -> Self {
-        Self { client }
+impl CallMiddleware {
+    pub fn new(client: &Arc<Client>) -> Self {
+        Self {
+            client: client.clone(),
+        }
     }
 }
 
 #[async_trait]
-impl Middleware<CallRequest, Result<JsonValue, Error>> for UpstreamMiddleware {
+impl Middleware for CallMiddleware {
     async fn call(
         &self,
-        request: CallRequest,
-        _next: NextFn<CallRequest, Result<JsonValue, Error>>,
+        request: Request,
+        next: NextFn<Request, Result<JsonValue, Error>>,
```

Comment:
this should be terminating. there shouldn't be any middleware after it.
we will just have another subscription middleware that only handles subscriptions